### ISSUE TITLE
Asset Naming Conventions

### ISSUE THREAD

**Issue by shawnrice** [shawnrice/alfred-bundler#36]
**TL;DR**: Assets names should be normalized in the JSON filenames as well as inside the JSON. The problem is that the `load` functions rely heavily on the `name` variable, keeping naming both rigid and inconsistent. Question: how to fix/should we fix?

---

The names of the assets as defined in the JSON defaults (and the JSON files) are fairly rigid and seem arbitrary.

The naming is strict and can resort to problems like #5; however, it's deeply integrated into the way that the Bundler locates and calls assets. So, `Terminal Notifier` is always referred to as `Terminal-Notifier`. The `-` was put in there because that's how it names itself, although the app is `terminal-notifier.app`. Cocoa Dialog is `cocoaDialog` because v2.1.1's application is called `cocoaDialog.app`; however, v 3.0.0-beta7 (current default) is `CocoaDialog.app`. The `invoke` file was supposed to help take care of these, and it does so but not the best way.

So, we should normalize these, at least through the bundler (and the bindings). Should we just make them consistent through each asset (i.e. make all versions of `cocoaDialog` become `CocoaDialog`), or should we make them consistent across all assets (i.e. all lowercase and spaces turn hyphenated, underscores, or camelCase)?

The more radical version would be to decouple the naming consistencies making them either case insensitive.

Thoughts?

**Comment by stephen-bunn** [comment]
It would probably be best to be as consistent as possible with asset names. Lowercase with hyphens sounds like a fairly good choice.

**Comment by shawnrice** [comment]
What about things like CFPropertyList or cocoaDialog? Should we lowercase all of CFPropertylist? Should we force the hyphenation of CocoaDialog (cocoa-dialog)?

We'll have to expand the installation instructions in each JSON file to include renaming any .app file. We shouldn't (and needn't) rename the internal binary (i.e. `cocoa-dialog.app/Contents/MacOS/CocoaDialog`).

**Comment by stephen-bunn** [comment]
Does changing the name of the resource file actually change the way the installation is performed? If so we should probably change that.

Now seeing that resource name you just posted I feel like camel case would be a better way to go. Unfortunately that's pretty much what we have right now. If you want me could start using bundle IDs for every resource.

**Comment by shawnrice** [comment]
The `load` functions return the asset paths. Each asset is installed at

```
data/assets/{type}/{name}/{version}/<<asset>>
```

In the same directory, we place a file called `invoke` that contains the rest of the path. So, the load function, for Terminal Notifier, looks for `invoke` in `data/assets/utility/Terminal-Notifier/{version}`. The invoke file for Terminal Notifier contains `terminal-notifier.app/Contents/MacOS/terminal-notifier`. So, the load function appends the contents of that file to the path and returns

```
{HOME}/Library/Application Support/Alfred 2/Workflow Data/alfred.bundler-{AB_VERSION}/data/assets/utility/Terminal-Notifier/{version}/terminal-notifier.app/Contents/MacOS/terminal-notifier
```

For things like CocoaDialog, one version is "cocoaDialog" while another is "CocoaDialog", but since they're the same asset, they should be in the same place. So, for here, we'd have to rename v2.1.1 from `cocoaDialog.app` to `CocoaDialog.app` because the former is what comes out of the zip file.

The installation instructions are there because we can't determine how each asset is packaged.

The bundle ids wouldn't be too good to use because we'd then have to create a special case for utilities rather than keeping them consistent.

**Comment by shawnrice** [comment]
If it helps show the weirdness that the bundler tries to account for, here's the `cocoaDialog.json` file, the `Workflows.json` file, and the `Pashua.json` file.

#### Cocoa Dialog

```
{
    "name": "cocoaDialog",
    "type": "utility",
    "versions": {
        "default": {
            "invoke": "cocoaDialog.app/Contents/MacOS/cocoaDialog",
            "files": [
                {
                    "url": "https://github.com/downloads/mstratman/cocoadialog/cocoaDialog_3.0.0-beta7.dmg",
                    "method": "direct"
                }
            ],
            "install": [
                "hdiutil attach -nobrowse -quiet '__FILE__'",
                "cp -fR /Volumes/cocoaDialog.app/CocoaDialog.app '__DATA__'",
                "hdiutil detach -quiet /Volumes/cocoaDialog.app"
            ]
        },
        "3.0.0-beta7": {
            "invoke": "CocoaDialog.app/Contents/MacOS/cocoaDialog",
            "files": [
                {
                    "url": "https://github.com/downloads/mstratman/cocoadialog/cocoaDialog_3.0.0-beta7.dmg",
                    "method": "direct"
                }
            ],
            "install": [
                "hdiutil attach -nobrowse -quiet '__FILE__'",
                "cp -fR /Volumes/cocoaDialog.app/CocoaDialog.app '__DATA__'",
                "hdiutil detach -quiet /Volumes/cocoaDialog.app"
            ]
        },
        "2.1.1": {
            "invoke": "cocoaDialog.app/Contents/MacOS/cocoaDialog",
            "files": [
                {
                    "url": "https://github.com/downloads/mstratman/cocoadialog/CocoaDialog-2.1.1.dmg",
                    "method": "direct"
                }
            ],
            "install": [
                "hdiutil attach -nobrowse -quiet '__FILE__'",
                "cp -fR /Volumes/CocoaDialog/CocoaDialog.app '__DATA__'",
                "hdiutil detach -quiet /Volumes/CocoaDialog"
            ]
        }
    }
}
```

#### Workflows (PHP library)

```
{
    "name": "Workflows",
    "type": "php",
    "versions": {
        "default": {
            "invoke": "workflows.php",
            "files": [
                {
                    "url": "https://raw.githubusercontent.com/jdfwarrior/Workflows/master/workflows.php",
                    "method": "direct"
                }
            ],
            "install": [
                "cp -f '__CACHE__/workflows.php' '__DATA__'"
            ]
        },
        "0.3.3": {
            "invoke": "workflows.php",
            "files": [
                {
                    "url": "https://raw.githubusercontent.com/jdfwarrior/Workflows/master/workflows.php",
                    "method": "direct"
                }
            ],
            "install": [
                "cp -f '__CACHE__/workflows.php' '__DATA__'"
            ]
        }
    }
}
```

#### Pashua

```
{
    "name": "Pashua",
    "type": "utility",
    "description": "show interactive dialogs",
    "versions": {
        "default": {
            "invoke": "Pashua.app/Contents/MacOS/Pashua",
            "files": [
                {
                    "url": "http://www.bluem.net/files/Pashua.dmg",
                    "method": "direct"
                }
            ],
            "install": [
                "hdiutil attach -nobrowse -quiet '__FILE__'",
                "cp -f /Volumes/Pashua/Documentation.html '__DATA__'",
                "cp -fR /Volumes/Pashua/Pashua.app '__DATA__'",
                "hdiutil detach -quiet /Volumes/Pashua"
            ]
        }
    }
}
```

These examples are relevant because you can see 
1. the differences in how Cocoa Dialog is named;
2. the ease of the Workflows installation / invocation (also, you can see a non-utility asset); and
3. the complexity of the Pashua installation.

**Comment by stephen-bunn** [comment]
Ah, I see now. How were you thinking of approaching this?

**Comment by shawnrice** [comment]
#### Related:

There isn't much here yet, but I've realized that, to facilitate development, I need to write out more about how the bundler works internally. So I'm trying to utilize the wiki for some [developer notes](https://github.com/shawnrice/alfred-bundler/wiki/Developer-Notes). There is [a page for the JSON](https://github.com/shawnrice/alfred-bundler/wiki/Asset-JSON-Structure) as well. Neither are well developed yet.

Below is a longer answer, bear with me.

#### Naming Conventions

##### Names that _might_ not matter

One problem with the bundler is that it's written in four languages, and so we're kind of using those naming conventions in each place. PHP filenames, for the most part, replace spaces with hyphens. The others replace spaces with underscores. 

Exceptions are 
- the actual bundler libraries (AlfredBundler.EXT); and
- the wrappers (alfred.bundler.EXT)

I figure that the one place file naming structures should be absolutely consistent are between the wrappers and the libraries. The reason is that it's easy enough to identify the main libraries and which files are associated with those. However, Python's implicit namespacing might mess with these, in which case, we should probably follow Python's naming requirements for filenames.

##### Names that **absolutely** matter

The file names that _really_ matter are the ones that the bundler uses as variables. The main variables that we need to worry about are the ones that come from the `load` and `icon` functions. Icons are easier, so let's deal with those first:

###### Icons

The icon directory structure is `data/assets/icons/{FONT}/{COLOR}/{NAME}`. The load order for icons, however, is `icon( font, name, color )`, which makes sense, but the information is more richly displayed in the file structure the way it is. Hence, those variables need to be normalized, which is pretty easy to do by:
1. Making everything lowercase; and
2. Changing 3-letter hex codes to 6-letter hex codes.

###### Regular Assets

In order to prevent really chaotic code, the paths have been standardized to be: `data/assets/{type}/{name}/{version}`. One nice aspect is that different versions of the same asset live right next to each other in obviously named folders. There are some current exceptions with PHP, Python, and Ruby in regards to Composer Packages, Packages from PyPi, and Gems — respectively, but those are necessary deviates due to how the composer/pip/gems function.

Regardless, PHP-only assets are in the `php` folder, and the same goes for the other languages. Theoretically, this structure can also support assets for any language: Applescript, Go, Groovy, etc.... The workflow author would just call the appropriate `load` function to install them and ignore the paths.

Hence, the `type` variable is set well enough, and the `version` variable is set well enough (as they should follow the name/case of the version released). The only real problem comes with the `name` variable.

#### Conceptual Problem with Naming

To simplify things, the [installation script](https://github.com/shawnrice/alfred-bundler/wiki/Developer-Notes%3A-Verbose#installation) uses only a JSON filepath and an optional version to install an asset. However, the `load` functions that call the installation script whittles down the arguments to decide which JSON to use via the `name` variable. So, sans custom JSON, the `load` functions look for the `default` JSON in `bundler/meta/defaults/{NAME}.json`. The `load` function uses the same `name` variable later to find the asset in the directory path. So, the `name` field in the JSON needs to match the final installation path.

Now, re-using this `name` variable for the identification of the defaults as well as the directory structure might have been a lazy shortcut, but it also saves us from having to read the JSON each time a default asset is loaded (after it has already been installed). But that means that a slight variation in the names spells failure.

Is this rigidity okay? Or should we have better ways to catch the error? (Again, I didn't think of it much until #5, which I solved with a band-aid).

This also presents at least a minor problem when the naming isn't consistent within different versions of an asset (cocoaDialog/CocoaDialog is the only default example).

But enforcing a strict lowercase/hyphen ultimately changes the name of the asset and might actually strip information about the asset, but keeping the name also might cause clashes. I.e. [CFPropertyList](https://github.com/rodneyrehm/CFPropertyList), the php library included as a default takes its name from Core Foundation Property List, tying its functionality to Apple's original C implementation of plists. Converting to lowercase removes the implied abbreviation (did I mention that I'm getting my PhD in Rhetoric? My anxiety over this issue might also be influenced by that). Name clashes here might also come up because there is a [Ruby CFPropertyList](http://rubyforge.org/projects/cfpropertylist/), a [Python CFPropertyList](https://github.com/bencochran/CFPropertyList), and probably others.

David Ferguson's Workflows library gets the `Workflows.json` filename, but, luckily for us, each [Alfred Library](http://www.alfredforum.com/topic/2030-workflow-libraries-and-helpers/) does have a different name, although all are, for good reasons, similar. But, if we wanted to specify these further, then we'd have to change 'Workflows.json' to 'workflows-php.json' or create subdirectories to organize the assets (`bundler/meta/defaults/php/workflows.php`). The `load` function would then be responsible for looking for the default in the correct sub-directory.

Another option would be to have a `defaults.json` file that would just have a map to each default asset.

Overall naming consistency is important for DX, especially to avoid things like PHP's fucking annoying `needle`-`haystack` inconsistencies:
- `strpos(haystack, needle)`
- `preg_match(pattern, subject)`

The workflow author shouldn't be left looking up whether the call is to `Terminal-Notifier` and the next to `cocoaDialog`. Right?

#### Installation Instructions

Rewriting the installation instructions shouldn't matter. Basically, if you look at those JSON files above, we'd just add in whichever lines were necessary to normalize the asset.

Again, sorry for the length.

**Comment by shawnrice** [comment]
Just to write this down somewhere:

Once we freeze the naming standards, we should make the bindings fit with it. Then, the bundlers should also include a way to load the bindings (i.e. for PHP something like `$object->binding( 'CocoaDialog' );`)

The bindings would then `require` or `include` the binding -- however the language needed them.

**Comment by deanishe** [comment]
> The workflow author shouldn't be left looking up whether the call is to Terminal-Notifier and the next to cocoaDialog. Right?

I don't have a big problem with this myself. Users will typically have to consult the documentation for the utility/library in question, unless they're intimately familiar with it, and in either case, asking the bundler for the utility/library in question by its given name seems a more obvious, less magical solution.

If we start enforcing our own naming conventions, that's just something else for users to learn.

> Name clashes here might also come up because there is a Ruby CFPropertyList, a Python CFPropertyList, and probably others.

How have you implemented the Ruby dependency bundler? This isn't an issue with the Python wrapper, as anything Python-specific is kept separate end-to-end (in the API and on disk).

**Comment by shawnrice** [comment]
@deanishe: You're right about changing naming conventions.

So, then should we worry about rectifying inconsistencies within a particular asset such as Cocoa Dialog?

For Ruby, the clash between a default and a Gem won't exist because the user needs to use the `gem` method to load one. The Python implementation is similar because it looks to the defaults only for utilities, right?

I know that the implementation of the "defaults" that ship with the bundler follow the logic of PHP and Bash better than with Python and Ruby, and the utilities _shouldn't_ clash. But are there cases in which a Python or Ruby developer would want to "bundle" something that wouldn't be available in PyPi or as a Gem? `CFPropertyList` is a poor example because it's a PHP library but then a PyPi package and a Gem, so there wouldn't be the clash there because we use different mechanisms. But are there cases when, say, somewhat writing in Ruby calls a general `load` function and gets a PHP library instead?

We could mitigate these clashes by subdividing the `defaults` directory to be:

```
bundler/meta/defaults/bash
bundler/meta/defaults/php
bundler/meta/defaults/python
bundler/meta/defaults/ruby
bundler/meta/defaults/utility
bundler/meta/defaults/applescript
...
```

But would that be worth it? I know this isn't really an issue _right now_, but it _could_ become one in the future.

**Comment by deanishe** [comment]
WRT cocoaDialog, I say we use the name its latest version has. Same goes for other utilities. Going forward, we should stick with the names they had at the time of the Taurus release. We can consider any renaming at the next major release.

Pip can also install from GitHub repos, so authors can keep any non PyPi stuff on there. 

I don't think divvying up the assets by language is a 100% solution: what if Rita and I each wrote, say, our own completely incompatible Keychain libraries for Python, both called "keychain"?

Until we can think of a more reliable naming convention (somehow URL based?), I think the standard, asset-like install shouldn't be used for Python libs.

**Comment by shawnrice** [comment]
We obviously need to keep the naming standards when we release Taurus because changing them would effectively break the API.

The problem is a bundler-wide issue and not just confined to the Python bundler because it potentially produces clashes between languages.

The clashes shouldn't be a problem when using Pip, Gem, and Composer because those already take care of it.

So, if both you and @Ritashugisha wrote a Python library called "keychain," and I wrote wrote a PHP one with the same name, then we'd have to sort out the namespace. The only real way to deal with it currently is that the first one that made it into the 'defaults' would claim the "keychain" "namespace" in the bundler, which means that you'd call that one by "keychain" where we might have to call another one "Keychain-PHP" and a last "Keychain-Dean".

Subdividing by language would make it so that my library didn't clash with yours and Rita's, which helps the problem slightly but doesn't solve it.

It also makes it so that we JSON in the defaults cannot be "extended" to include other versions (again: #5).

Excluding the Python bundler from using the defaults just puts the issue off but doesn't get rid of it. Currently, that would be the issue that would force **Gemini** to come into existence (although there is something poetic about a branch called Gemini having its _raison d'être_ to be resolving twinning conflicts). And, even though major version changes are exciting (**burn it down so we can rebuild entirely!**), they should be strongly avoided as supporting multiple major versions has negative effects on development and, by extension, adoption.

**Comment by shawnrice** [comment]
Actually, to better focus the issue:

The problem is two-fold in that one is an internal issue and the other is a user ("external") issue.

The user problem comes from the arguments needed for the generic `load` function.

The internal problem comes from the way we use the current directory structure as a shortcut, or, perhaps better said, as an address.

**Comment by deanishe** [comment]
> The problem is a bundler-wide issue and not just confined to the Python bundler because it potentially produces clashes between languages.
> 
> The clashes shouldn't be a problem when using Pip, Gem, and Composer because those already take care of it.

If pip, gem and composer take care of it, it's not a problem because there's only bash left… :wink:

Seriously, the obvious solution to my mind is to do what everyone else does, including Alfred, Apple and Java, and require the name of custom assets to be something like `com.example.my-alfred-asset`.

Requiring long-ass names for utilities/libraries is not ideal, but it's probably the best way to avoid clashes, and should be familiar to any Alfred developer (bundle IDs).

**Comment by shawnrice** [comment]
Well, composer takes care of only some of it. Not too many people use composer when writing workflows. Usually, they just load some libraries in the regular "download/require_once" way.

Pip, Gem, and Composer don't matter because they don't have any `defaults` that are recognized by the Bundler. So, the important ones here are any assets not loaded through Pip, Gem, or Composer. For the Python version, those are usually just utilities. For PHP there could be more. And I'm not yet sure about Ruby.

So, maybe we do need to do the naming scheme that is something closer to a bundleid.

**Comment by shawnrice** [comment]
Actually, here's a good question: for Python, how can we get the bundler to work with
1. @deanishe : your Python Library;
2. Alp; and
3. Alfred-Python?

It's not through Pip, right? So we'd need to include them in the defaults.

**Comment by shawnrice** [comment]
Are there any more thoughts for this, or should we just leave the naming scheme alone for Taurus and close this issue?

**Comment by deanishe** [comment]
**TL;DR** Adding the bundlet to `Alfred-Workflow` is a better idea than adding `Alfred-Workflow` to the bundler (though that's possible). Forget `Alp` and `Alfred-Python`.

---

None of them can currently be installed via pip (they don't have a `setup.py` file).

I can add one to `Alfred-Workflow` (it's something I should have done long ago, tbh). We could also create `setup.py` files for the others and send pull requests. Then `pip` can install them from GitHub.

They could all be installed fairly easily via the defaults mechanism. (`Alfred-Workflow` in particular includes a zip file that contains just the actual library without all the tests and other chaff.)

I'm blowing pretty hard on my own trumpet here, but I would recommend against including `Alp` and/or `Alfred-Python`. Neither has been updated in over a year (excluding my update to `Alfred-Python`), and so they don't support more recent Alfred features, including custom subtexts, environmental variables and `arg` as a tag, not an attribute (whatever the actual advantage of that might be).

`Alfred-Workflow` was very late to the party, and is in many respects a "next-gen" library compared to the other two (due to my blatant theft of all the good ideas from other libraries and avoidance of the iffy ones). It has many features they don't, does most of the same things better, and has a saner interface.

`Alp` has a couple of advantages with support for notifications and its bundled `requests` library pisses on my `web.py` from a great height, but these are entirely moot in the context of the bunder, which provides way better support for notifications and the ability to install a recent version of `requests` (`Alp` bundles an obsolete 1.x version; `requests` is currently at 2.x. The API is different.)

So, while we _could_ include any or all of them as defaults, I think only my very own `Alfred-Workflow` should be considered for inclusion.

But that said, I'd recommend using `Alfred-Workflow` to include the bundler, instead of the other way around.

It has the totally cool feature (which I totally stole) of catching errors, logging them and showing them to the user, which is to say, if you import the bundler from within `Alfred-Workflow` you'll see an error in Alfred if something fucks up, rather than Alfred's characteristic utter silence.

I also designed it to be small: 106 KB as source code, 26 KB as a zip file, which you can `import` and use as long as you don't need the `background.py` module.

**Comment by shawnrice** [comment]
I'm fine not adding those in. Yours is definitely more curated and does have many advantages (I can see this from not knowing much of any Python).

The "which to include first" problem between Alfred-Workflows and the bundler is interesting, but I don't think that we should necessarily make it one way or another. You could copy the Python bundlet into Alfred-Workflows, but I'd suggest making sure that there is a way to avoid any sort of clash if the user wants to load the library via the bundler instead.

For instance, I was writing a "re-install the bundler" test for the PHP bundler, and I just ended up wrapping each class with an "if ( ! class_exists( classname ) ) ..." in order to avoid the error when redeclaring classes. Maybe there could be some functionality like that. Not sure.

Could you write up the JSON to make it so that the bundler can install your library? I'm assigning you to that to finish whenever.

**Comment by deanishe** [comment]
Thinking about this a bit more, I think it's a bad idea:

The Python bundler uses `pip` to install Python libraries. They are installed by the `bundler.init()` function and all go in the same, workflow-specific directory. We've discussed at length why this is the only practical way to support Python libraries (the hidden, recursive dependencies many libraries have).

Any Python libraries installed by the bundler would be installed by a different function (presumably a new one called `bundler.library()`) and would go in the assets directory.

It's no great technical challenge, but that would leave the Python bundler with two different functions for installing Python libraries, and the difference isn't obvious to a user: why do they have to install `requests` using `requirements.txt` and `bundler.init()`, but `Alfred-Workflow` via `bundler.library('Alfred-Workflow')`?

It'd be possible to beef-up `bundler.init()` to also handle installing libraries via the standard bundler API, but I think that's just not worth the effort and additional complexity (for us and users) when we're currently only talking about a couple of Python libraries.

I think it'd be a much better idea for me to put `Alfred-Workflow` on PyPi where it can be installed like every other library.

**Comment by shawnrice** [comment]
That makes it simple. Put it on PyPi and the issue becomes moot for that sort of thing. 

Otherwise, the dual functionality (or other) would need to exist in order to download/copy/symlink Alfred-Workflow into the workflow specific folder.

The first is definitely the path of least resistance.

So, I'm going to close this.
